loader/loader.go

```diff
@@ -47,114 +54,188 @@ var (
 		{"cdnjs", cdnjs, regexp.MustCompile(`^cdnjs.com/libraries/([^/]+)(?:/([(\d\.)]+-?[^/]*))?(?:/(.*))?$`)},
 		{"github", github, regexp.MustCompile(`^github.com/([^/]+)/([^/]+)/(.*)$`)},
 	}
-	invalidScriptErrMsg = `The file "%[1]s" couldn't be found on local disk, ` +
-		`and trying to retrieve it from https://%[1]s failed as well. Make ` +
-		`sure that you've specified the right path to the file. If you're ` +
+	httpsSchemeCouldntBeLoadedMsg = `The moduleSpecifier "%s" couldn't be retrieved from` +
+		` the resolved url "%s". Error : "%s"`
+	fileSchemeCouldntBeLoadedMsg = `The moduleSpecifier "%s" couldn't be found on ` +
+		`local disk. Make sure that you've specified the right path to the file. If you're ` +
 		`running k6 using the Docker image make sure you have mounted the ` +
 		`local directory (-v /local/path/:/inside/docker/path) containing ` +
 		`your script and modules so that they're accessible by k6 from ` +
 		`inside of the container, see ` +
 		`https://docs.k6.io/v1.0/docs/modules#section-using-local-modules-with-docker.`
+	errNoLoaderMatched = errors.New("no loader matched")
 )
 
-// Resolves a relative path to an absolute one.
-func Resolve(pwd, name string) string {
-	if name[0] == '.' {
-		return filepath.ToSlash(filepath.Join(pwd, name))
+// Resolve a relative path to an absolute one.
+func Resolve(pwd *url.URL, moduleSpecifier string) (*url.URL, error) {
+	if moduleSpecifier == "" {
+		return nil, errors.New("local or remote path required")
 	}
-	return name
-}
 
-// Returns the directory for the path.
-func Dir(name string) string {
-	if name == "-" {
-		return "/"
+	// we always want for the pwd to end in a slash, but filepath/path.Clean strips it so we readd
+	// it if it's missing
+	if pwd.Opaque != "" {
+		if !strings.HasSuffix(pwd.Opaque, "/") {
+			pwd.Opaque += "/"
+		}
+	} else if !strings.HasSuffix(pwd.Path, "/") {
+		pwd.Path += "/"
```

Comment:
It is very, very strange for this to be here - if I call `Resolve(pwd, myFile)`, I won't expect for `pwd` to be changed, but that's exactly what this code would do...

Reply:
I agree it's strange but:
1. there should always be a slash at the end of the pwd otherwise this function will not work properly when we resolve
2.  as is written in the comment above (I suppose I moved it up when doing adding the above if) filepath/path.Clean removes the slash ... for some reason and we do call such methods NUMEROUS times through the code. This is even more prevelant in the tests and a lot of filepath/path function call Clean in of themselves so if you ever do something to Pwd's path with them you need to add the slash back. 
All of this combined finally let to me deciding that there is no point in doing this part at every place where this is done so I just do it here.
Arguably we can do on a copy of the pwd just before pwd.Parse, but to be honest I prefer if I fix the pwd either way

Reply:
In my head, if this code has to be here, then it *needs* to be on a copy of the `pwd`. Having a function called `Resolve` that is called like `realPath, err := Resolve(pwd, myFile)`, and having the seemingly side-effect free function change one of its arguments is breaking all kinds of expectations...